src/state-visualizer/visualizer.py

```diff
@@ -0,0 +1,141 @@
+import graphviz
+import time
+from PyQt5.QtWidgets import *
+from PyQt5.QtCore import *
+from PyQt5.QtGui import QPainter
+from PyQt5.QtSvg import QSvgRenderer
+
+import rospy
+from smach_msgs.msg import SmachContainerStatus,SmachContainerStructure
+from threading import Lock
+from dataclasses import dataclass, field
+from typing import Optional
+
+@dataclass
+class State:
+    name : str
+    children : list #of states
+
+@dataclass
+class StateMachine:
+    states : dict  = field(default_factory=dict)#maps string name to State
+    structure : Optional[SmachContainerStructure] = None
+    mutex : Lock = Lock()
+    cur_active : str = ""
+    needs_redraw : bool = True
+    def set_active_state(self, active_state):
+        """
+        sets the state specified to be active
+        """
+        if active_state != self.cur_active and active_state in self.states:
+            self.cur_active = active_state
+            self.needs_redraw = True
+
+    def rebuild(self, structure : SmachContainerStructure):
+        """
+        rebuilds the state dictionary with a new structure message
+        """
+        self.states = {child : State(child, []) for child in structure.children}
+        for start, end in zip(structure.outcomes_from, structure.outcomes_to):
+            if end != "None":
+                self.states[start].children.append(self.states[end])
+        self.needs_redraw = True
+
+
+    def check_rebuild(self, structure : SmachContainerStructure):
+        """
+        checks if the structure passed as input matches the structure already represented
+        """
+        if structure == self.structure:
+            return False
+        else:
+            self.rebuild(structure)
+
+stateMachine = StateMachine()
+
+def container_status_callback(status : SmachContainerStatus):
+    #rospy.loginfo(status.active_states)
+    stateMachine.mutex.acquire()
```

Comment:
I think you can use `with` here

e.g. `with stateMachine.mutex:`

also as per our style guide it should be `state_machine`